## This is a Demo of the model used in wealthsmart application. <br> It includes contents such as implementing collaborative filtering and calculating cosine similarity.

### step 1: Load user data

In [30]:
import pandas as pd

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

user_data = pd.read_csv("/Users/kayleecho/Documents/data science projects/recommendation_system/synthetic_user_data.csv", sep=",")
print(user_data.shape)
user_data.head(10)
user_data

(1000, 23)


user_id  investment_goal_long_term_growth  investment_goal_stable_income  \
0          1                              0.29                           0.24   
1          2                              0.04                           0.38   
2          3                              0.39                           0.02   
3          4                              0.32                           0.12   
4          5                              0.18                           0.18   
..       ...                               ...                            ...   
995      996                              0.45                           0.07   
996      997                              0.29                           0.31   
997      998                              0.32                           0.18   
998      999                              0.24                           0.27   
999     1000                              0.08                           0.26   

     investment_goal_balanced_approach  investment_goal_high_risk_high_reward  \
0                                 0.29                                   0.14   
1                                 0.06                                   0.33   
2                                 0.15                                   0.17   
3                                 0.30                                   0.03   
4                                 0.01                                   0.27   
..                                 ...                                    ...   
995                               0.20                                   0.25   
996                               0.16                                   0.24   
997                               0.34                                   0.05   
998                               0.12                                   0.05   
999                               0.06                                   0.21   

     investment_goal_short_term_trading  risk_tolerance  \
0                                  0.04            4.83   
1                                  0.19            5.04   
2                                  0.27            7.28   
3                                  0.23            2.97   
4                                  0.36            5.61   
..                                  ...             ...   
995                                0.03            4.81   
996                                0.00            6.01   
997                                0.11            1.74   
998                                0.32            7.01   
999                                0.39            2.44   

     holding_period_response  holding_period_reaction_strength  \
0                       3.33                              7.39   
1                       6.67                              8.58   
2                       3.33                              1.53   
3                      10.00                              6.59   
4                       1.00                              5.57   
..                       ...                               ...   
995                     3.33                              6.34   
996                     6.67                              3.75   
997                     3.33                              2.85   
998                    10.00                              5.56   
999                     6.67                              5.35   

     expected_return_score  product_preference_stocks  \
0                     3.15                      18.72   
1                     4.23                      16.37   
2                     2.09                      12.67   
3                     7.55                      11.44   
4                     9.24                      11.17   
..                     ...                        ...   
995                   9.19                      22.61   
996                   6.00                       3.98   
997                   5.65                  

### Split the user data into two parts. <br> new users from user_id = 1 to user_id = 20, so total of 20 users <br> old users from user_id = 20 to user_id = 1000, so total of 980 users

In [59]:
new_users = user_data[0:20]
old_users = user_data[20:]

print(user_data.columns)
print(new_users.shape)
print(old_users.shape)

Index(['user_id', 'investment_goal_long_term_growth',
       'investment_goal_stable_income', 'investment_goal_balanced_approach',
       'investment_goal_high_risk_high_reward',
       'investment_goal_short_term_trading', 'risk_tolerance',
       'holding_period_response', 'holding_period_reaction_strength',
       'expected_return_score', 'product_preference_stocks',
       'product_preference_etfs', 'product_preference_government_bonds',
       'product_preference_corporate_bonds',
       'product_preference_municipal_bonds',
       'product_preference_savings_bonds', 'product_preference_derivatives',
       'product_preference_alternative_investments',
       'factor_weight_investment_goals', 'factor_weight_risk_tolerance',
       'factor_weight_holding_period', 'factor_weight_return_score',
       'factor_weight_product_preference'],
      dtype='object')
(20, 23)
(980, 23)


### Let's Try our second method first which is using multidimensional similarity measures.
* we first drop the coulumns that we don't need. There are 17 columns that we will consider when calculating score for this method.
* this includes, investment goal + risk tolerance + holding period(composite) + return score + product preference.

In [32]:
new_users = new_users.drop(['user_id', 'factor_weight_investment_goals', 'factor_weight_risk_tolerance', 'factor_weight_holding_period', 'factor_weight_return_score', 'factor_weight_product_preference'], axis=1)
old_users = old_users.drop(['user_id', 'factor_weight_investment_goals', 'factor_weight_risk_tolerance', 'factor_weight_holding_period', 'factor_weight_return_score', 'factor_weight_product_preference'], axis=1)
print(new_users.columns)
print(old_users.columns)
print(new_users.shape)
print(old_users.shape)

Index(['investment_goal_long_term_growth', 'investment_goal_stable_income',
       'investment_goal_balanced_approach',
       'investment_goal_high_risk_high_reward',
       'investment_goal_short_term_trading', 'risk_tolerance',
       'holding_period_response', 'holding_period_reaction_strength',
       'expected_return_score', 'product_preference_stocks',
       'product_preference_etfs', 'product_preference_government_bonds',
       'product_preference_corporate_bonds',
       'product_preference_municipal_bonds',
       'product_preference_savings_bonds', 'product_preference_derivatives',
       'product_preference_alternative_investments'],
      dtype='object')
Index(['investment_goal_long_term_growth', 'investment_goal_stable_income',
       'investment_goal_balanced_approach',
       'investment_goal_high_risk_high_reward',
       'investment_goal_short_term_trading', 'risk_tolerance',
       'holding_period_response', 'holding_period_reaction_strength',
       'expected_retu

#### Just to show you how cosine similarity score calculation works in numpy. <br>This is not related to our project.

In [33]:
import numpy as np
from numpy.linalg import norm

A = np.array([0.40, 0.10, 0.40, 0.05, 0.05, 0.75, 0.667, 0.700, 0.85, 0.40, 0.20, 0.15, 0.05, 0.05, 0.10, 0.05])
B = np.array([0.70, 0.10, 0.10, 0.05, 0.05, 0.625, 1.00, 0.85, 0.725, 0.50, 0.10, 0.10, 0.05, 0.05, 0.10, 0.10])
C = np.array([0.10, 0.10, 0.20, 0.50, 0.10, 0.875, 0.333, 0.50, 0.90, 0.30, 0.15, 0.15, 0.10, 0.10, 0.15, 0.15])

print("A:", A)
print("B:", B)
print("C:", C)
 
# compute cosine similarity
cosine_ab = np.dot(A,B)/(norm(A)*norm(B))
cosine_ac = np.dot(A,C)/(norm(A)*norm(C))
print("Cosine Similarity:", cosine_ab)
print("Cosine Similarity:", cosine_ac)

A: [0.4   0.1   0.4   0.05  0.05  0.75  0.667 0.7   0.85  0.4   0.2   0.15
 0.05  0.05  0.1   0.05 ]
B: [0.7   0.1   0.1   0.05  0.05  0.625 1.    0.85  0.725 0.5   0.1   0.1
 0.05  0.05  0.1   0.1  ]
C: [0.1   0.1   0.2   0.5   0.1   0.875 0.333 0.5   0.9   0.3   0.15  0.15
 0.1   0.1   0.15  0.15 ]
Cosine Similarity: 0.9459992837365002
Cosine Similarity: 0.9000536145871202


#### calculate cosine similarity for our user dataset. <br> columns: old user, row: new user (from user_1 to user20)

In [34]:
import numpy as np
from numpy.linalg import norm
import pandas as pd

new_users = new_users.to_numpy()
old_users = old_users.to_numpy()

cosine_similarities = {}

for i in range(len(new_users)):
    similarities = []
    for j in range(len(old_users)):
        cosine_similarity = np.dot(new_users[i], old_users[j]) / (norm(new_users[i]) * norm(old_users[j]))
        similarities.append(cosine_similarity)
    cosine_similarities[f"User_{i+1}"] = similarities

cosine_sim_df = pd.DataFrame.from_dict(cosine_similarities, orient='index')

print(cosine_sim_df.shape)
cosine_sim_df

(20, 980)


0         1         2         3         4         5         6    \
User_1   0.646780  0.705349  0.885670  0.750899  0.542893  0.928692  0.797425   
User_2   0.664223  0.671111  0.894914  0.783612  0.652424  0.907447  0.786290   
User_3   0.774081  0.818284  0.803140  0.864367  0.786401  0.775371  0.758335   
User_4   0.818190  0.674707  0.882047  0.760602  0.604008  0.824650  0.683368   
User_5   0.720922  0.648949  0.723963  0.874438  0.833017  0.625708  0.734087   
User_6   0.785634  0.824023  0.886742  0.863545  0.797613  0.852791  0.739012   
User_7   0.663771  0.723641  0.852763  0.843971  0.688338  0.852205  0.710485   
User_8   0.752609  0.766952  0.874620  0.846135  0.698280  0.826294  0.733952   
User_9   0.789602  0.684952  0.880462  0.818131  0.739923  0.790223  0.591353   
User_10  0.705911  0.716362  0.742157  0.733189  0.674235  0.667933  0.818196   
User_11  0.818278  0.858438  0.867034  0.874588  0.747880  0.834530  0.697879   
User_12  0.687851  0.525258  0.763681  0.641478  0.586974  0.669687  0.680818   
User_13  0.555404  0.575670  0.689442  0.768573  0.724484  0.681647  0.641523   
User_14  0.728254  0.677768  0.689013  0.893258  0.785845  0.667190  0.609356   
User_15  0.651525  0.777368  0.919893  0.766802  0.701217  0.777400  0.774203   
User_16  0.731310  0.822406  0.888417  0.786467  0.684224  0.890334  0.679388   
User_17  0.694780  0.766487  0.877043  0.785472  0.598024  0.898279  0.840431   
User_18  0.412389  0.761880  0.680343  0.678239  0.566282  0.783856  0.845837   
User_19  0.560348  0.728751  0.696366  0.831159  0.795748  0.770933  0.658452   
User_20  0.696776  0.588295  0.592317  0.847323  0.814597  0.478274  0.568295   

              7         8         9         10        11        12        13   \
User_1   0.665439  0.821752  0.794411  0.686605  0.755189  0.851475  0.664331   
User_2   0.785012  0.895936  0.865743  0.834901  0.737822  0.779236  0.802265   
User_3   0.648232  0.804735  0.835167  0.731491  0.758994  0.659659  0.757574   
User_4   0.577341  0.870128  0.746053  0.823371  0.816324  0.771760  0.841124   
User_5   0.739374  0.800152  0.862352  0.708530  0.815716  0.512007  0.795167   
User_6   0.800314  0.958357  0.913389  0.865372  0.775412  0.766700  0.859608   
User_7   0.747662  0.732859  0.801265  0.630357  0.807925  0.841848  0.781800   
User_8   0.670639  0.766704  0.771540  0.686889  0.845872  0.814681  0.811469   
User_9   0.748049  0.866880  0.813015  0.765968  0.850438  0.793711  0.897592   
User_10  0.514900  0.835188  0.759043  0.867568  0.597081  0.447311  0.726296   
User_11  0.768379  0.892652  0.873032  0.760400  0.847910  0.855519  0.820605   
User_12  0.559170  0.897005  0.718147  0.876249  0.657366  0.504591  0.758808   
User_13  0.696645  0.635250  0.768600  0.568570  0.663487  0.557349  0.673960   
User_14  0.741818  0.684567  0.813794  0.518520  0.892030  0.698210  0.729498   
User_15  0.650444  0.798438  0.737886  0.769788  0.690275  0.698618  0.832341   
User_16  0.790103  0.908225  0.872660  0.844966  0.692534  0.849360  0.812467   
User_17  0.610379  0.807538  0.790320  0.719105  0.742950  0.766019  0.699068   
User_18  0.667469  0.690708  0.797757  0.625617  0.423305  0.610985  0.440794   
User_19  0.947637  0.782126  0.926494  0.675832  0.645370  0.695979  0.704119   
User_20  0.707073  0.650491  0.766069  0.501856  0.857027  0.452656  0.739456   

              14        15        16        17        18        19        20   \
User_1   0.886615  0.747447  0.739509  0.938726  0.789263  0.825181  0.849470   
User_2   0.862583  0.896952  0.749236  0.944459  0.830105  0.916997  0.939644   
User_3   0.800871  0.688260  0.637617  0.898745  0.628894  0.814977  0.831254   
User_4   0.796798  0.785610  0.762816  0.939779  0.718733  0.828458  0.808405   
User_5   0.805772  0.905861  0.627856  0.834548  0.617649  0.785093  0.792051   
User_6   0.809191  0.786575  0.772184  0.899716  0.777154  0.963620  0.906090   
User_7   0.719131  0.819456  0.777383  0.88

#### find transpose of the matrix above in order to sort them in descending order. <br> we are interested in those that have high cosine similarity score.

In [60]:
cosine_sim_df_transposed = cosine_sim_df.T
print(cosine_sim_df_transposed.shape)
cosine_sim_df_transposed.head(10)

(980, 20)


User_1  User_2  User_3  User_4  User_5  User_6  User_7  User_8  User_9  \
0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
1     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
2     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
3     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
4     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
5     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
6     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
7     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
8     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   
9     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     1.0   

   User_10  User_11  User_12  User_13  User_14  User_15  User_16  User_17  \
0      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
1      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
2      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
3      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
4      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
5      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
6      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
7      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
8      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   
9      1.0      1.0      1.0      1.0      1.0      1.0      1.0      1.0   

   User_18  User_19  User_20  
0      1.0      1.0      1.0  
1      1.0      1.0      1.0  
2      1.0      1.0      1.0  
3      1.0      1.0      1.0  
4      1.0      1.0      1.0  
5      1.0      1.0      1.0  
6      1.0      1.0      1.0  
7      1.0      1.0      1.0  
8      1.0      1.0      1.0  
9      1.0      1.0      1.0

#### Match the new user to the old user with highest cosine similarity score. <br> Sort them in descending order (standard should be cosine similarity score)

In [61]:
cosine_sim_sorted = pd.DataFrame()

for i in cosine_sim_df_transposed.columns:
    sorted_indices = cosine_sim_df_transposed[i].sort_values(ascending=False).index
    cosine_sim_sorted[i] = cosine_sim_df_transposed[i].sort_values(ascending=False).values
    cosine_sim_sorted["old_user_id_for_" + i] = cosine_sim_df_transposed.index[sorted_indices].values

cosine_sim_sorted

User_1  old_user_id_for_User_1  User_2  old_user_id_for_User_2  User_3  \
0       1.0                       0     1.0                       0     1.0   
1       1.0                     644     1.0                     644     1.0   
2       1.0                     646     1.0                     646     1.0   
3       1.0                     647     1.0                     647     1.0   
4       1.0                     648     1.0                     648     1.0   
..      ...                     ...     ...                     ...     ...   
975     1.0                     330     1.0                     330     1.0   
976     1.0                     331     1.0                     331     1.0   
977     1.0                     332     1.0                     332     1.0   
978     1.0                     333     1.0                     333     1.0   
979     1.0                     979     1.0                     979     1.0   

     old_user_id_for_User_3  User_4  old_user_id_for_User_4  User_5  \
0                         0     1.0                       0     1.0   
1                       644     1.0                     644     1.0   
2                       646     1.0                     646     1.0   
3                       647     1.0                     647     1.0   
4                       648     1.0                     648     1.0   
..                      ...     ...                     ...     ...   
975                     330     1.0                     330     1.0   
976                     331     1.0                     331     1.0   
977                     332     1.0                     332     1.0   
978                     333     1.0                     333     1.0   
979                     979     1.0                     979     1.0   

     old_user_id_for_User_5  User_6  old_user_id_for_User_6  User_7  \
0                         0     1.0                       0     1.0   
1                       644     1.0                     644     1.0   
2                       646     1.0                     646     1.0   
3                       647     1.0                     647     1.0   
4                       648     1.0                     648     1.0   
..                      ...     ...                     ...     ...   
975                     330     1.0                     330     1.0   
976                     331     1.0                     331     1.0   
977                     332     1.0                     332     1.0   
978                     333     1.0                     333     1.0   
979                     979     1.0                     979     1.0   

     old_user_id_for_User_7  User_8  old_user_id_for_User_8  User_9  \
0                         0     1.0                       0     1.0   
1                       644     1.0                     644     1.0   
2                       646     1.0                     646     1.0   
3                       647     1.0                     647     1.0   
4                       648     1.0                     648     1.0   
..                      ...     ...                     ...     ...   
975                     330     1.0                     330     1.0   
976                     331     1.0                     331     1.0   
977                     332     1.0                     332     1.0   
978                     333     1.0                     333     1.0   
979                     979     1.0                     979     1.0   

     old_user_id_for_User_9  User_10  old_user_id_for_User_10  User_11  \
0                         0      1.0                        0      1.0   
1                       644      1.0                      644      1.0   
2                       646      1.0                      646      1.0   
3                       647      1.0                      647      1.0   
4                       648      1.0                      648      1.0   
..                      ...      ...                 

### Let's Try our first method which is using the modified weighted composite score. <br> We specify this to two different mini methods.
#### 1. calculating cosine similarity using five vector space instead of using one composite score. <br> => turns out this first method doesn't work because in 1 dimensional array cosine similarity is just 1 for everyone… I will explain this further later. <br> 2. using one composite score by adding all five weighted vector space.

#### Load User Data Again!

In [37]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 100)
user_data = pd.read_csv("/Users/kayleecho/Documents/data science projects/recommendation_system/synthetic_user_data.csv", sep=",")

print(user_data.shape)
user_data.head(10)

(1000, 23)


user_id  investment_goal_long_term_growth  investment_goal_stable_income  \
0        1                              0.29                           0.24   
1        2                              0.04                           0.38   
2        3                              0.39                           0.02   
3        4                              0.32                           0.12   
4        5                              0.18                           0.18   
5        6                              0.11                           0.43   
6        7                              0.09                           0.14   
7        8                              0.34                           0.15   
8        9                              0.01                           0.11   
9       10                              0.18                           0.37   

   investment_goal_balanced_approach  investment_goal_high_risk_high_reward  \
0                               0.29                                   0.14   
1                               0.06                                   0.33   
2                               0.15                                   0.17   
3                               0.30                                   0.03   
4                               0.01                                   0.27   
5                               0.13                                   0.02   
6                               0.26                                   0.28   
7                               0.21                                   0.29   
8                               0.22                                   0.41   
9                               0.26                                   0.16   

   investment_goal_short_term_trading  risk_tolerance  \
0                                0.04            4.83   
1                                0.19            5.04   
2                                0.27            7.28   
3                                0.23            2.97   
4                                0.36            5.61   
5                                0.31            2.27   
6                                0.23            1.32   
7                                0.01            8.80   
8                                0.25            5.55   
9                                0.03            3.06   

   holding_period_response  holding_period_reaction_strength  \
0                     3.33                              7.39   
1                     6.67                              8.58   
2                     3.33                              1.53   
3                    10.00                              6.59   
4                     1.00                              5.57   
5                     1.00                              1.54   
6                     6.67                              6.57   
7                    10.00                              1.02   
8                     3.33                              3.81   
9                     6.67                              5.90   

   expected_return_score  product_preference_stocks  product_preference_etfs  \
0                   3.15                      18.72                    17.62   
1                   4.23                      16.37                    15.63   
2                   2.09                      12.67                    20.59   
3                   7.55                      11.44                     8.32   
4                   9.24                      11.17                     8.25   
5                   2.43                      12.85                    15.70   
6                   4.58                       6.13                    15.85   
7                   3.78                       9.20                    10.97   
8                   3.95                       3.92                     9.29   
9                   9.87                      26.20                     9.83   

   product_preference_government_bonds  product_p

#### Normalize all the data that are created using random_allocation function in sythetic_user_data.py
Remember description of the function? <br>
This function generates a list of n random numbers that sum to total of 100.
Used when generating product preferences and factor weights.
#### We do the normalization by dividing the value by 100.

In [38]:
# the weights normally should add up to 1, so we are dividing each weight by 100 to match the scale!
w1 = user_data['factor_weight_investment_goals'] / 100  # Investment Goals weight
w2 = user_data['factor_weight_risk_tolerance'] / 100    # Risk Tolerance weight
w3 = user_data['factor_weight_holding_period'] / 100    # Holding Period weight
w4 = user_data['factor_weight_return_score'] / 100      # Return Score weight
w5 = user_data['factor_weight_product_preference'] / 100 # Product Preference weight

# if we are to use composite score, product preference score should also be in similar range compared to other categories in the composite score.
user_data["product_preference_stocks"] = user_data["product_preference_stocks"] / 100
user_data["product_preference_etfs"] = user_data["product_preference_etfs"] / 100
user_data["product_preference_government_bonds"] = user_data["product_preference_government_bonds"] / 100
user_data["product_preference_corporate_bonds"] = user_data["product_preference_corporate_bonds"] / 100
user_data["product_preference_municipal_bonds"] = user_data["product_preference_municipal_bonds"] / 100
user_data["product_preference_savings_bonds"] = user_data["product_preference_savings_bonds"] / 100
user_data["product_preference_derivatives"] = user_data["product_preference_derivatives"] / 100
user_data["product_preference_alternative_investments"] = user_data["product_preference_alternative_investments"] / 100

### Below is the score function we use.
> Score = w1 * Investment Goals + w2 * Risk Tolerance + w3 * Holding Period + w4 * Return Score (Expected) + w5 * Product Preference

In [39]:
# Investment goals score
investment_goal_cols = [
    'investment_goal_long_term_growth',
    'investment_goal_stable_income',
    'investment_goal_balanced_approach',
    'investment_goal_high_risk_high_reward',
    'investment_goal_short_term_trading'
]

# Risk tolerance score
risk_tolerance_score = user_data['risk_tolerance']

# Holding period score
holding_period_cols = [
    'holding_period_response',
    'holding_period_reaction_strength'
]

# Expected return score
return_score = user_data['expected_return_score']

# Product preference score
product_preference_cols = [
    'product_preference_stocks',
    'product_preference_etfs',
    'product_preference_government_bonds',
    'product_preference_corporate_bonds',
    'product_preference_municipal_bonds',
    'product_preference_savings_bonds',
    'product_preference_derivatives',
    'product_preference_alternative_investments'
]

### For those that have more than one score to consider, we create composite score for each category before multiplying weight and adding it to the larger composite score calculation. The weights multiplied below follow this logic.

#### Investment Goal Composite Score Weight: <br>
Discrete Option (Categorical Selection): <br>
1: Long term growth <br>
1: Stable Income <br>
2: Balanced Approach <br>
3: High-Risk, High-Reward <br>
3: Short Term Trading & Speculation <br>
1, 2, 3 are in groups where they tend to be more similar to each other <br>
1 = (Lower-risk or steady-growth focus) <br>
2 = (Medium risk and time horizon) <br>
3 = (Higher volatility, aggressive approach) <br>

#### Holding Period Composite Score Weight: <br>
Question we are going to ask to the customers: Imagine you started with a $10,000 investment. Then, in one month, your investment lost $1,000 in value. What would you do next? <br> 
1. I’d sell everything  2. I’d sell some   3. I’d do nothing    4. I’d buy more <br>
Mapping to a Decimal Scale:  <br>
Option 1: I’d sell everything: 1.00  <br>
Option 2: I’d sell some: ~3.33  <br>
Option 3: I’d do nothing: ~6.67  <br>
Option 4: I’d buy more: 10.00 (genuinely risky, higher volatility & aggressive)  <br>
Holding period reaction:  <br>
1 - 10: the higher the number,  customers are willing to take risks and the smaller the number becomes they become risk averse.  <br>
So technically, when creating composite scores for this section, we are going to add holding_period_response and holding_period_reaction_strength because the higher the score of the sum it would indicate that the customer tends to take risks and the smaller the sum score is they tend to be risk averse.

#### Product Preference Composite Score: <br>
Ask users to distribute a total of 100 points(or $100) among the product categories. <br>
Type: Discrete allocation with decimals allowed (e.g. 12.5 points) <br>
Constraint: Sum must equal 100 <br>
Conversion: we normalized these to sum to 1.00 above. <br>
Available Products Lists: <br>
Equities <br>
Stocks, ETFs <br>
Government Bonds <br>
Corporate Bonds <br>
Derivatives (파생상품) <br>
Alternative Investments <br>

Maybe we can add these as we proceed. Not in the current data and model calcualtion though. <br>
Real Estate Investment Trusts <br>
NFTs <br>
Cryptocurrencies <br>

In [62]:
# For those that have more than one score to consider, we create composite score for each category before multiplying weight and adding it to the larger composite score calculation
investment_goal_composite_score = 1 * user_data["investment_goal_long_term_growth"]  + 1 * user_data["investment_goal_stable_income"] + 2 * user_data["investment_goal_balanced_approach"] + 3 * user_data["investment_goal_high_risk_high_reward"] + 3 * user_data["investment_goal_short_term_trading"]

holding_period_composite_score = user_data["holding_period_response"] + user_data["holding_period_reaction_strength"]

product_preference_composite_score = 1 * user_data["product_preference_government_bonds"] + 2.5 * user_data["product_preference_corporate_bonds"] \
                                    + 4 * user_data["product_preference_stocks"] + 4 * user_data["product_preference_etfs"] \
                                    + 6 * user_data["product_preference_derivatives"] + 8 * user_data["product_preference_alternative_investments"]

#### 1. calculating cosine similarity using five vector space instead of using one composite score.

In [44]:
# This is calculating similarity using five vector space instead of using one composite score!
import numpy as np
np.set_printoptions(threshold=np.inf)

import numpy as np

score_vector = np.column_stack((
    w1 * investment_goal_composite_score,
    w2 * risk_tolerance_score,
    w3 * holding_period_composite_score,
    w4 * return_score,
    w5 * product_preference_composite_score
))

print(score_vector.shape)
score_vector[326]

(1000, 5)


array([0.525915 , 0.936888 , 3.198633 , 0.062244 , 0.6038832])

In [45]:
from numpy.linalg import norm

new_users = score_vector[:20]
old_users = score_vector[20:]

cosine_similarities = {}

for i in range(len(new_users)):
    similarities = []
    for j in range(len(old_users)):
        cosine_similarity = np.dot(new_users[i], old_users[j]) / (norm(new_users[i]) * norm(old_users[j]))
        similarities.append(cosine_similarity)
    cosine_similarities[f"User_{i+1}"] = similarities

cosine_sim_df = pd.DataFrame.from_dict(cosine_similarities, orient='index')

print(cosine_sim_df.shape)
cosine_sim_df

(20, 980)


0         1         2         3         4         5         6    \
User_1   0.870835  0.821462  0.500509  0.990703  0.404884  0.925333  0.395929   
User_2   0.920025  0.806379  0.651027  0.972784  0.328799  0.966035  0.569226   
User_3   0.792190  0.912482  0.565602  0.889011  0.540157  0.916863  0.416367   
User_4   0.966096  0.719777  0.622587  0.951738  0.210625  0.952319  0.623591   
User_5   0.697033  0.724107  0.991898  0.471380  0.551556  0.676771  0.948929   
User_6   0.678080  0.785113  0.961718  0.567165  0.544381  0.742709  0.834193   
User_7   0.831406  0.773666  0.474402  0.919195  0.247600  0.943718  0.395518   
User_8   0.486068  0.789705  0.693150  0.481260  0.895110  0.497761  0.505248   
User_9   0.387331  0.667368  0.833259  0.234770  0.827360  0.369842  0.698947   
User_10  0.649524  0.752528  0.988619  0.438818  0.648086  0.638785  0.916165   
User_11  0.767711  0.860072  0.770695  0.724402  0.736194  0.747668  0.660975   
User_12  0.549863  0.893480  0.683212  0.576804  0.892683  0.626161  0.476452   
User_13  0.477584  0.879904  0.590218  0.605007  0.920570  0.578577  0.322704   
User_14  0.829319  0.861629  0.512894  0.911984  0.524475  0.885174  0.408100   
User_15  0.667137  0.973702  0.721400  0.753329  0.852344  0.773213  0.486543   
User_16  0.350374  0.801253  0.859577  0.300640  0.884642  0.468300  0.613562   
User_17  0.909740  0.790789  0.536745  0.984873  0.373829  0.922183  0.469928   
User_18  0.680554  0.723966  0.905622  0.414564  0.535886  0.673651  0.906467   
User_19  0.734376  0.970677  0.668848  0.873516  0.749765  0.850332  0.434970   
User_20  0.780789  0.639961  0.921307  0.505228  0.434561  0.678232  0.972034   

              7         8         9         10        11        12        13   \
User_1   0.398468  0.972754  0.795985  0.375756  0.947416  0.960808  0.994079   
User_2   0.321126  0.960718  0.761873  0.404027  0.966429  0.877512  0.940146   
User_3   0.500931  0.913263  0.868353  0.588353  0.855682  0.944967  0.886223   
User_4   0.243721  0.962193  0.658369  0.242078  0.976434  0.879842  0.929785   
User_5   0.625231  0.527308  0.660411  0.578272  0.489199  0.451255  0.483852   
User_6   0.543712  0.586776  0.734299  0.703276  0.569608  0.474260  0.532167   
User_7   0.192306  0.947675  0.706226  0.411359  0.942843  0.920030  0.866115   
User_8   0.962575  0.463003  0.818139  0.627491  0.345990  0.531513  0.591924   
User_9   0.916291  0.253096  0.669891  0.639955  0.150583  0.277183  0.327165   
User_10  0.722435  0.490740  0.700669  0.633307  0.433116  0.441166  0.469305   
User_11  0.819466  0.728658  0.849201  0.520927  0.643416  0.756108  0.808249   
User_12  0.917357  0.586225  0.897539  0.723803  0.464118  0.675590  0.661265   
User_13  0.918180  0.565511  0.918700  0.734594  0.448293  0.648029  0.688188   
User_14  0.535531  0.926106  0.829981  0.434883  0.862933  0.975388  0.947471   
User_15  0.844928  0.738182  0.977453  0.761761  0.641677  0.781675  0.803862   
User_16  0.871892  0.309873  0.799896  0.905792  0.214722  0.313833  0.331491   
User_17  0.397477  0.970526  0.761395  0.307316  0.950561  0.946563  0.997729   
User_18  0.613399  0.532402  0.628872  0.557812  0.465305  0.532308  0.440771   
User_19  0.721564  0.838475  0.975217  0.704859  0.767670  0.851495  0.899202   
User_20  0.565443  0.575110  0.567057  0.352166  0.541381  0.513932  0.545738   

              14        15        16        17        18        19        20   \
User_1   0.719632  0.727031  0.942337  0.911788  0.494428  0.791166  0.514668   
User_2   0.769647  0.876559  0.989441  0.973869  0.335386  0.881487  0.515640   
User_3   0.704972  0.670107  0.863964  0.813098  0.594939  0.828978  0.598171   
User_4   0.762122  0.889653  0.985916  0.995432  0.264464  0.793920  0.434652   
User_5   0.906260  0.898206  0.676477  0.662501  0.473603  0.781063  0.786905   
User_6   0.841094  0.899950  0.733096  0.700478  0.388695  0.903243  0.752877   
User_7   0.574666  0.694417  0.895529  0.87

#### 1. find transpose of the matrix above in order to sort them in descending order. <br> 2. we are interested in those that have high cosine similarity score. <br> 3. Match the new user to the old user with highest cosine similarity score. <br> 4. Sort them in descending order (standard should be cosine similarity score)

In [63]:
cosine_sim_df_transposed = cosine_sim_df.T

cosine_sim_sorted = pd.DataFrame()

for i in cosine_sim_df_transposed.columns:
    sorted_indices = cosine_sim_df_transposed[i].sort_values(ascending=False).index
    cosine_sim_sorted[i] = cosine_sim_df_transposed[i].sort_values(ascending=False).values
    cosine_sim_sorted["old_user_id_for_" + i] = cosine_sim_df_transposed.index[sorted_indices].values

cosine_sim_sorted

User_1  old_user_id_for_User_1  User_2  old_user_id_for_User_2  User_3  \
0       1.0                       0     1.0                       0     1.0   
1       1.0                     644     1.0                     644     1.0   
2       1.0                     646     1.0                     646     1.0   
3       1.0                     647     1.0                     647     1.0   
4       1.0                     648     1.0                     648     1.0   
..      ...                     ...     ...                     ...     ...   
975     1.0                     330     1.0                     330     1.0   
976     1.0                     331     1.0                     331     1.0   
977     1.0                     332     1.0                     332     1.0   
978     1.0                     333     1.0                     333     1.0   
979     1.0                     979     1.0                     979     1.0   

     old_user_id_for_User_3  User_4  old_user_id_for_User_4  User_5  \
0                         0     1.0                       0     1.0   
1                       644     1.0                     644     1.0   
2                       646     1.0                     646     1.0   
3                       647     1.0                     647     1.0   
4                       648     1.0                     648     1.0   
..                      ...     ...                     ...     ...   
975                     330     1.0                     330     1.0   
976                     331     1.0                     331     1.0   
977                     332     1.0                     332     1.0   
978                     333     1.0                     333     1.0   
979                     979     1.0                     979     1.0   

     old_user_id_for_User_5  User_6  old_user_id_for_User_6  User_7  \
0                         0     1.0                       0     1.0   
1                       644     1.0                     644     1.0   
2                       646     1.0                     646     1.0   
3                       647     1.0                     647     1.0   
4                       648     1.0                     648     1.0   
..                      ...     ...                     ...     ...   
975                     330     1.0                     330     1.0   
976                     331     1.0                     331     1.0   
977                     332     1.0                     332     1.0   
978                     333     1.0                     333     1.0   
979                     979     1.0                     979     1.0   

     old_user_id_for_User_7  User_8  old_user_id_for_User_8  User_9  \
0                         0     1.0                       0     1.0   
1                       644     1.0                     644     1.0   
2                       646     1.0                     646     1.0   
3                       647     1.0                     647     1.0   
4                       648     1.0                     648     1.0   
..                      ...     ...                     ...     ...   
975                     330     1.0                     330     1.0   
976                     331     1.0                     331     1.0   
977                     332     1.0                     332     1.0   
978                     333     1.0                     333     1.0   
979                     979     1.0                     979     1.0   

     old_user_id_for_User_9  User_10  old_user_id_for_User_10  User_11  \
0                         0      1.0                        0      1.0   
1                       644      1.0                      644      1.0   
2                       646      1.0                      646      1.0   
3                       647      1.0                      647      1.0   
4                       648      1.0                      648      1.0   
..                      ...      ...                 

#### 2. Using one composite score by adding all five weighted vector space.


In [48]:
# Calculate one composite score!
composite_score = np.column_stack((
    w1 * investment_goal_composite_score + 
    w2 * risk_tolerance_score + 
    w3 * holding_period_composite_score + 
    w4 * return_score + 
    w5 * product_preference_composite_score 
))

composite_score


array([[ 5.21254025,  6.07971592,  3.57250075,  8.2698788 ,  5.152695  ,
         2.6148735 ,  4.10410577,  6.03729025,  4.30834332,  4.7289821 ,
         6.37867713,  2.85633308,  4.97721179,  3.36280781,  3.78086906,
         4.7605574 ,  6.14813212,  4.66634077,  3.74915752,  5.62599192,
         4.31151056,  3.85159492,  3.96246835,  6.14096624,  5.5283501 ,
         3.94531862,  4.83538989,  6.32539605,  5.6045844 ,  5.23407248,
         3.66286852,  4.89128602,  4.98844505,  7.2972434 ,  7.56792018,
         5.23987168,  4.66773972,  7.30225259,  4.67560808,  4.23176924,
         5.08035768,  6.59256063,  6.32141375,  5.76457853,  5.53877697,
         3.41601263,  4.54713368,  7.42049827,  5.14633554,  6.62555272,
         3.86856105,  5.47167393,  5.39351044,  6.0370312 , 10.16755445,
         3.59350276,  5.80388516,  3.7393728 ,  6.84655875,  5.7211988 ,
         3.38657958,  5.63391752,  4.18551264,  6.32402692,  3.72600632,
         3.80874954,  3.57882364,  5.548203  ,  6.0

In [49]:
from numpy.linalg import norm

composite_score = composite_score.T

new_users = composite_score[:20]
old_users = composite_score[20:]

cosine_similarities = {}

for i in range(len(new_users)):
    similarities = []
    for j in range(len(old_users)):
        cosine_similarity = np.dot(new_users[i], old_users[j]) / (norm(new_users[i]) * norm(old_users[j]))
        similarities.append(cosine_similarity)
    cosine_similarities[f"User_{i+1}"] = similarities

cosine_sim_df = pd.DataFrame.from_dict(cosine_similarities, orient='index')

print(cosine_sim_df.shape)
cosine_sim_df


(20, 980)


0    1    2    3    4    5    6    7    8    9    10   11   12   13   \
User_1   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_2   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_3   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_4   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_5   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_6   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_7   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_8   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_9   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_10  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_11  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_12  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_13  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_14  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_15  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_16  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_17  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_18  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_19  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_20  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   

         14   15   16   17   18   19   20   21   22   23   24   25   26   27   \
User_1   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_2   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_3   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_4   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_5   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_6   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_7   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_8   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_9   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_10  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_11  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_12  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_13  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_14  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_15  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_16  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_17  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_18  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_19  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_20  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   

         28   29   30   31   32   33   34   35   36   37   38   39   40   41   \
User_1   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_2   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_3   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_4   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_5   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_6   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
User_7   1.0  1.0  1.0  1.0  1.0  1.0 

#### When there is only one dimensional vector space, cosine similarity becomes trival because it will cancel out to one.. Hence, if all our composite scores are positive, which is exactly our case, we will get 1.0 for every pair! So we are going to use different metric here. Just measuring the "Distance" is what we're going to use. <br> This will measure the straight line distance between the two vectors.

In [56]:
import numpy as np
import pandas as pd

new_users = composite_score[:20]
old_users = composite_score[20:]

result_distances = {}

for i in range(len(new_users)):
    distances = []
    for j in range(len(old_users)):
        distance_array = np.abs(new_users[i] - old_users[j])
        distance = float(distance_array.item())
        disstance = round(distance, 5)
        distances.append(distance)
    result_distances[f"User_{i+1}"] = distances

result_dist_df = pd.DataFrame.from_dict(result_distances, orient='index')

print("distances shape:", result_dist_df.shape)
result_dist_df

distances shape: (20, 980)


0         1         2         3         4         5         6    \
User_1   0.901030  1.360945  1.250072  0.928426  0.315810  1.267222  0.377150   
User_2   1.768205  2.228121  2.117248  0.061250  0.551366  2.134397  1.244326   
User_3   0.739010  0.279094  0.389968  2.568465  1.955849  0.372818  1.262889   
User_4   3.958368  4.418284  4.307410  2.128913  2.741529  4.324560  3.434489   
User_5   0.841184  1.301100  1.190227  0.988271  0.375655  1.207376  0.317305   
User_6   1.696637  1.236721  1.347595  3.526093  2.913477  1.330445  2.220516   
User_7   0.207405  0.252511  0.141637  2.036860  1.424244  0.158787  0.731284   
User_8   1.725780  2.185695  2.074822  0.103676  0.508940  2.091972  1.201900   
User_9   0.003167  0.456748  0.345875  1.832623  1.220007  0.363025  0.527047   
User_10  0.417472  0.877387  0.766514  1.411984  0.799368  0.783663  0.106408   
User_11  2.067167  2.527082  2.416209  0.237711  0.850327  2.433359  1.543287   
User_12  1.455177  0.995262  1.106135  3.284633  2.672017  1.088986  1.979057   
User_13  0.665701  1.125617  1.014743  1.163754  0.551138  1.031893  0.141822   
User_14  0.948703  0.488787  0.599661  2.778158  2.165542  0.582511  1.472582   
User_15  0.530642  0.070726  0.181599  2.360097  1.747481  0.164450  1.054521   
User_16  0.449047  0.908962  0.798089  1.380409  0.767793  0.815239  0.074832   
User_17  1.836622  2.296537  2.185664  0.007166  0.619782  2.202814  1.312742   
User_18  0.354830  0.814746  0.703872  1.474625  0.862009  0.721022  0.169049   
User_19  0.562353  0.102437  0.213311  2.391809  1.779193  0.196161  1.086232   
User_20  1.314481  1.774397  1.663524  0.514974  0.097642  1.680673  0.790602   

              7         8         9         10        11        12        13   \
User_1   1.112856  0.392044  0.021532  1.549672  0.321254  0.224095  2.084703   
User_2   0.245680  0.475132  0.845643  2.416847  1.188430  1.091271  1.217527   
User_3   2.752895  2.032084  1.661572  0.090368  1.318785  1.415944  3.724743   
User_4   1.944483  2.665294  3.035806  4.607010  3.378593  3.281434  0.972635   
User_5   1.172701  0.451889  0.081377  1.489826  0.261409  0.164250  2.144548   
User_6   3.710523  2.989711  2.619199  1.047995  2.276413  2.373572  4.682370   
User_7   2.221290  1.500479  1.129967  0.441237  0.787180  0.884339  3.193138   
User_8   0.288106  0.432706  0.803218  2.374422  1.146004  1.048845  1.259953   
User_9   2.017053  1.296241  0.925729  0.645475  0.582943  0.680102  2.988900   
User_10  1.596414  0.875602  0.505090  1.066114  0.162304  0.259463  2.568261   
User_11  0.053281  0.774093  1.144605  2.715809  1.487391  1.390232  0.918566   
User_12  3.469063  2.748251  2.377739  0.806535  2.034953  2.132112  4.440910   
User_13  1.348184  0.627373  0.256861  1.314343  0.085926  0.011233  2.320032   
User_14  2.962588  2.241777  1.871265  0.300061  1.528478  1.625637  3.934436   
User_15  2.544527  1.823715  1.453203  0.118001  1.110417  1.207576  3.516374   
User_16  1.564839  0.844027  0.473515  1.097689  0.130729  0.227888  2.536686   
User_17  0.177264  0.543548  0.914060  2.485264  1.256846  1.159687  1.149111   
User_18  1.659055  0.938244  0.567732  1.003472  0.224945  0.322104  2.630903   
User_19  2.576239  1.855427  1.484915  0.086289  1.142128  1.239288  3.548086   
User_20  0.699404  0.021408  0.391919  1.963123  0.734706  0.637547  1.671251   

              14        15        16        17        18        19        20   \
User_1   2.355380  0.027331  0.544801  2.089712  0.536932  0.980771  0.132183   
User_2   1.488204  0.839844  1.411976  1.222537  1.404108  1.847947  0.999358   
User_3   3.995419  1.667371  1.095239  3.729752  1.103107  0.659268  1.507857   
User_4   0.701959  3.030007  3.602139  0.967626  3.594271  4.038110  3.189521   
User_5   2.415225  0.087177  0.484955  2.149558  0.477087  0.920926  0.072337   
User_6   4.953047  2.624998  2.052866  4.687379  2.060735  1.616896  2.465484   
User_7   3.463814  1.135766  0.563634  3.19

In [58]:
result_dist_df_transposed = result_dist_df.T

dist_sorted = pd.DataFrame()

for i in result_dist_df_transposed.columns:
    sorted_indices = result_dist_df_transposed[i].sort_values(ascending=True).index
    dist_sorted[i] = result_dist_df_transposed[i].sort_values(ascending=True).values
    dist_sorted["old_user_id_for_" + i] = result_dist_df_transposed.index[sorted_indices].values 

dist_sorted

User_1  old_user_id_for_User_1    User_2  old_user_id_for_User_2  \
0    0.000551                     570  0.001047                     692   
1    0.000823                     175  0.002142                     221   
2    0.000949                     877  0.002619                     768   
3    0.004161                     813  0.007315                     829   
4    0.005935                     314  0.007348                      77   
..        ...                     ...       ...                     ...   
975  4.955014                      34  4.218221                      65   
976  5.085397                      65  4.237964                     705   
977  5.105140                     705  4.306199                     232   
978  5.300183                     148  4.433007                     148   
979  5.743709                     327  4.876533                     327   

       User_3  old_user_id_for_User_3    User_4  old_user_id_for_User_4  \
0    0.003098                     209  0.002589                     276   
1    0.003914                     850  0.027036                      78   
2    0.006323                      46  0.035885                     914   
3    0.008080                     973  0.042952                     302   
4    0.008500                     822  0.059518                     978   
..        ...                     ...       ...                     ...   
975  6.595054                      34  5.910117                     949   
976  6.725436                      65  5.942131                     428   
977  6.745179                     705  6.094760                     682   
978  6.940222                     148  6.253462                     312   
979  7.383748                     327  6.496362                     232   

       User_5  old_user_id_for_User_5    User_6  old_user_id_for_User_6  \
0    0.000395                     886  0.001108                     884   
1    0.001288                     292  0.046172                     781   
2    0.002545                     395  0.048634                      62   
3    0.004591                     628  0.053010                     960   
4    0.006359                      28  0.053724                     880   
..        ...                     ...       ...                     ...   
975  5.014859                      34  7.552681                      34   
976  5.145242                      65  7.683064                      65   
977  5.164985                     705  7.702806                     705   
978  5.360028                     148  7.897850                     148   
979  5.803554                     327  8.341376                     327   

       User_7  old_user_id_for_User_7    User_8  old_user_id_for_User_8  \
0    0.000035                     345  0.000259                      33   
1    0.000440                     318  0.000882                     192   
2    0.001933                     708  0.001888                     557   
3    0.003319                     378  0.002699                     957   
4    0.006669                     854  0.005036                     811   
..        ...                     ...       ...                     ...   
975  6.063449                      34  4.260647                      65   
976  6.193831                      65  4.263773                     232   
977  6.213574                     705  4.280390                     705   
978  6.408617                     148  4.475433                     148   
979  6.852143                     327  4.918959                     327   

       User_9  old_user_id_for_User_9   User_10  old_user_id_for_User_10  \
0    0.000195                     736  0.003287                      116   
1    0.003167                       0  0.003734                      512   
2    0.003390                     817  0.005774                      679   
3    0.004598                     109  0.006668                       80   
4    0.006558          